pkg/applier/stackapplier.go

```diff
@@ -55,57 +55,60 @@ func (s *StackApplier) Run(ctx context.Context) error {
 		return nil // The context is already done.
 	}
 
+	trigger := make(chan struct{}, 1)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
 		return fmt.Errorf("failed to create watcher: %w", err)
 	}
-	defer watcher.Close()
-
-	debounceCtx, cancelDebouncer := context.WithCancel(ctx)
-	defer cancelDebouncer()
-
-	debouncer := debounce.Debouncer[fsnotify.Event]{
-		Input:    watcher.Events,
-		Timeout:  1 * time.Second,
-		Filter:   s.triggersApply,
-		Callback: func(fsnotify.Event) { s.apply(debounceCtx) },
-	}
-
-	// Send an artificial event to ensure that an initial apply will happen.
-	go func() { watcher.Events <- fsnotify.Event{} }()
 
-	// Consume and log any errors.
 	go func() {
-		for {
-			err, ok := <-watcher.Errors
-			if !ok {
-				return
-			}
-			s.log.WithError(err).Error("Error while watching stack")
-		}
+		err = s.runWatcher(watcher, trigger, ctx.Done())
```

Comment:
I don’t think `err =` is needed now

Reply:
It is, if we want to forward any errors that might occur when running the watcher to the caller. But: we have to close the trigger channel _after_ assigning to `err` here. Therefore, at least that defer must be pulled one level up from `runWatcher`.

Reply:
Not sure if the errCh would need to be buffered and have it closed in the defer of the runWatcher function though.

Reply:
The main purpose of this PR is to decouple the stack application from the watcher loop, to ensure that long-running stack applications do not create backpressure into the kernel inotify event queue. While I agree your proposed change is more straight forward, it's quite important to not block the event consumption.

Reply:
That's a good point, I went a bit far in my simplification. I think if you used the channel approach but kept the triggers that would make the non-obvious coupling between when to close the triggers and the `err = ` being populated by the time the program returns.